pkg/feeds/sequence/sequence.go

```diff
@@ -80,89 +94,110 @@ func NewAsyncFinder(getter storage.Getter, feed *feeds.Feed) feeds.Lookup {
+	return &path{
+		base:   p.index,
+		index:  p.index,
+		max:    p.level,
+		level:  p.level,
+		chunk:  p.chunk,
+		cancel: make(chan struct{}),
+	}
+}
 func newPath(base uint64) *path {
-	return &path{base: base, cancel: make(chan struct{})}
+	return &path{base: base, cancel: make(chan struct{}), level: DefaultLevels, max: DefaultLevels}
 }
 
+// results capture a chunk lookup on a path
 type result struct {
-	chunk swarm.Chunk
-	path  *path
-	level int
-	seq   uint64
-	diff  int64
+	chunk swarm.Chunk // the chunk found
+	path  *path       // for a request
+	level int         // the
+	index uint64      // the actual seqeuence in seq
 }
 
 // At looks up the version valid at time `at`
 // after is a unix time hint of the latest known update
 func (f *asyncFinder) At(ctx context.Context, at, after int64) (ch swarm.Chunk, cur, next feeds.Index, err error) {
-	ch, diff, err := f.get(ctx, at, 0)
+	ch, err = f.get(ctx, at, 0)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 	if ch == nil {
-		return nil, nil, nil, nil
-	}
-	if diff == 0 {
-		return ch, &index{0}, &index{1}, nil
+		return nil, nil, &index{0}, nil
 	}
 	c := make(chan result)
-	p := newPath(0)
-	p.latest.chunk = ch
-	for p.level = 1; diff>>p.level > 0; p.level++ {
-	}
+	pa := newPath(0)
+	pa.chunk = ch
 	quit := make(chan struct{})
 	defer close(quit)
-	go f.at(ctx, at, p, c, quit)
+	go f.at(ctx, at, pa, c, quit)
 	for r := range c {
-		p = r.path
+		// r.path ~ tagged which path it comes from
+		// collect the results into the path
+		p := r.path
 		if r.chunk == nil {
-			if r.level == 0 {
-				return p.latest.chunk, &index{p.latest.seq}, &index{p.latest.seq + 1}, nil
-			}
-			if p.level < r.level {
+			if p.max < r.level {
 				continue
 			}
-			p.level = r.level - 1
+			p.max = r.level - 1
 		} else {
-			if r.diff == 0 {
-				return r.chunk, &index{r.seq}, &index{r.seq + 1}, nil
-			}
-			if p.latest.level > r.level {
+			p.close()
+			if p.min > r.level { // ignore lower than
 				continue
 			}
-			p.close()
-			p.latest = r
+			// if there is a chunk for this path, then the  latest chunk has surely beed sent
+			// since `at` starts from log interval
+			if p.level == r.level && r.level < DefaultLevels {
+				return r.chunk, &index{r.index}, &index{r.index + 1}, nil
+			}
+			p.min = r.level
+			p.chunk = r.chunk
+			p.index = r.index
 		}
-		// below applies even  if  p.latest==maxLevel
-		if p.latest.level == p.level {
-			if p.level == 0 {
-				return p.latest.chunk, &index{p.latest.seq}, &index{p.latest.seq + 1}, nil
+		// below applies even  if  p.latest==maxLevel in which case we just continue with
+		// DefaultLevel lookaheads
+		if p.min == p.max {
+			if p.min == 0 {
+				return p.chunk, &index{p.index}, &index{p.index + 1}, nil
 			}
-			p.close()
-			np := newPath(p.latest.seq)
-			np.level = p.level
-			np.latest.chunk = p.latest.chunk
+			np := p.next()
+			go f.at(ctx, at, np, c, quit)
+		}
+		if p.max < p.min {
+			np := newPath(p.base)
```

Comment:
this is incorrect. we should just use p.Next() and reset max and level to p.level